simulation/test_torrent_status.cpp

```diff
@@ -481,7 +481,7 @@ TORRENT_TEST(active_timer_no_seed)
 				// some part of the simulation is not deterministic, and causes this to vary
 				// between platforms/compilers
 				TEST_CHECK(active_time >= 10);
-				TEST_CHECK(active_time <= 14);
+				TEST_CHECK(active_time <= 15);
```

Comment:
This failed test seems to be one that expects to complete in ~14s, but now takes ~15s.

After some digging, it looks like at several stages do_ledbat() on the swarm peers sets `m_cwnd` to 0 due to excessive latency. It must then wait until the next timeout (1s) before its able to resume sending.

It's not uncommon for that to happen - it can be seen at multiple points in both the patched and unpatched test's `utp.log`. The test result logfiles become out of sync after a while when the unpatched one begins timing out and re-sending packets that the patched one no longer does, and by chance `m_cwnd` resetting to 0 appears to trigger one more time with this patch than without, delaying the test completing.

While pouring over the debug logs to figure that out I noted 12 lost packets and 5 resends in the unpatched test, though none in the patched one. So I guess that shows that it seems to be working as intended.

Reply:
you mean that `m_cwnd` is set to 1 MSS, right? (i.e. one packet size, not 0). That's supposed to happen when the send direction has become idle.

Reply:
https://github.com/arvidn/libtorrent/blob/b3d0148e641b94597c8ef132083e2226f9b0cefb/src/utp_stream.cpp#L3093-L3102

If latency is large enough to drop `m_cwnd` below 0, it looks like do_ledbat() sets it to 0. It then gets reset to the MSS on the next timeout (~1s). This wait for the timout appears to be whats causing the test to take 1s longer to complete.

From the BEP:
> If max_window becomes less than 0, it is set to 0. A window size of zero means that the socket may not send any packets. In this state, the socket will trigger a timeout and force the window size to one packet size, and send one packet. See the section on timeouts for more information.

Here it is in action. Note the first log message is printed just before its set to 0. The timestamps are microseconds:
```
[000013680048] 0x55f9ab26b940: do_ledbat delay:348978 off_target: -248978 window_factor:0.500000 target_factor:-2.489777 scaled_gain:-3734.664917 cwnd:2435 slow_start:0
[000013680048] 0x55f9ab26b940: no space in window send_buffer_size:423 cwnd:0 adv_wnd:1048576 in-flight:1451 mtu:1471
[000013680048] 0x55f9ab26b940: skipping send seq_nr:12652 ack_nr:30607 id:23865 target:50.0.0.1:6881 header_size:20 error:Success send_buffer_size:423 cwnd:0 adv_wnd:1048576 in-flight:1451 mtu:1471 effective-mtu:1471
...
[000013859928] 0x55f9ab26b940: do_ledbat delay:169458 off_target: -69458 window_factor:1.000000 target_factor:-0.694565 scaled_gain:-2083.694458 cwnd:0 slow_start:0
[000013859928] 0x55f9ab26b940: no space in window send_buffer_size:423 cwnd:0 adv_wnd:1048576 in-flight:0 mtu:1471
[000013859928] 0x55f9ab26b940: skipping send seq_nr:12652 ack_nr:30607 id:23865 target:50.0.0.1:6881 header_size:20 error:Success send_buffer_size:423 cwnd:0 adv_wnd:1048576 in-flight:0 mtu:1471 effective-mtu:1471
...
[000014500000] 0x55f9ab26b940: timeout num-timeouts: 0 max-resends: 3 confirmed: 1  acked-seq-num: 12651 mtu-seq: 0
[000014500000] 0x55f9ab26b940: resetting cwnd:1471
[000014500000] 0x55f9ab26b940: slow_start -> 1
[000014500000] 0x55f9ab26b940: sending packet seq_nr:12652 ack_nr:30607 type:ST_DATA id:23865 target:50.0.0.1:6881 size:443 error:Success send_buffer_size:0 cwnd:1471 adv_wnd:1048576 in-flight:0 mtu:1471 timestamp:15000000 time_diff:65120 mtu_probe:0 extension:0
[000014500000] 0x55f9ab26b940: updating timeout to: now + 500
```

Interestingly enough, I can't find this in the [LEDBAT rfc](https://datatracker.ietf.org/doc/rfc6817/) itself. There's a `MIN_CWND` floor with a suggested value of 2 (ie a 2*MSS floor) which it cannot fall below.

For curiosity I changed do_ledbat() to not lower `m_cwnd` below `m_mtu`, and that test now completes in 9.8s.

Reply:
that sounds like a very reasonable change (to clamp to 2 MSS)